protocols/mdns/src/behaviour.rs

```diff
@@ -72,19 +120,19 @@ pub struct GenMdns<S, T> {
     /// Future that fires when the TTL of at least one node in `discovered_nodes` expires.
     ///
     /// `None` if `discovered_nodes` is empty.
-    closest_expiration: Option<T>,
+    closest_expiration: Option<P::Timer>,
 }
 
-impl<S, T> GenMdns<S, T>
+impl<P> Behaviour<P>
 where
-    T: Builder,
+    P: Provider,
+    P::Timer: Builder,
```

Comment:
I think these could be specified on the trait itself, then you wouldn't have to repeat them here.